JAMES-3788 Implement SMTP proxy support for incoming connections usin…

ouvtam opened the pull request:
This PR enables proxy support for incoming SMTP connections. Since the implementation is on Netty level as suggested by chibenwa, it can also be used in IMAP/POP.

Hint: if you enable proxy support in smtpserver.xml and initiate a connection without "PROXY TCP4 ..." an exception is raised, thus the option is called proxyRequired. This might change in future.

Unfortunately, I was unable to run all tests on my Apple Silicon machine. I guess there are currently some limitations.

chibenwa commented on protocols/netty/pom.xml:
```diff
@@ -50,6 +50,11 @@
+        <dependency>
+            <groupId>io.netty</groupId>
+            <artifactId>netty-codec-haproxy</artifactId>
+            <version>4.1.72.Final</version>
```
```suggestion
            <version>${netty.version}</version>
```

ouvtam replied:
I will fix it.

chibenwa commented on protocols/netty/src/main/java/org/apache/james/protocols/netty/BasicChannelInboundHandler.java:
```diff
@@ -158,6 +163,21 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
+                    pSession.setProxyDestinationAddress(new InetSocketAddress(haproxyMsg.destinationAddress(), haproxyMsg.destinationPort()));
+                    pSession.setProxySourceAddress(new InetSocketAddress(haproxyMsg.sourceAddress(), haproxyMsg.sourcePort()));
+                } else {
+                    // TODO how to handle?
```
Maybe just throw an IllegalArgumentException ?

ouvtam replied:
Will do so.

chibenwa commented on protocols/smtp/src/test/java/org/apache/james/protocols/smtp/netty/NettyProxySMTPServerTest.java:
```diff
@@ -0,0 +1,135 @@
+package org.apache.james.protocols.smtp.netty;
```
The ASF V2 license header is compulsory.

ouvtam replied:
Of course.

chibenwa commented on protocols/api/src/main/java/org/apache/james/protocols/api/ProtocolSession.java:
```diff
@@ -148,10 +148,39 @@ public String toString() {
      */
     InetSocketAddress getRemoteAddress();
 
+    /**
+     * Return true if PROXY is required by the configuration
+     *
+     * @return supported
+     */
+    boolean isProxyRequired();
```
Doing a search on `isProxyRequired()` in this PR I do not see where this method is called, except in tests. Can we get rid of it?

ouvtam replied:
Seems legit.

chibenwa commented on protocols/api/src/main/java/org/apache/james/protocols/api/ProtocolSession.java:
```diff
@@ -148,10 +148,39 @@ public String toString() {
+     */
+    boolean isProxyRequired();
+
+    /**
+     * Return the {@link InetSocketAddress} of the proxy peer or null if proxy support is disabled.
+     */
+    InetSocketAddress getProxyDestinationAddress();
+
+    /**
+     * Sets the {@link InetSocketAddress} of the proxy peer.
+     */
+    void setProxyDestinationAddress(InetSocketAddress addr);
+
+    /**
+     * Return the {@link InetSocketAddress} of the remote peer if proxy is enabled or null if proxy
+     * support is disabled.
+     */
+    InetSocketAddress getProxySourceAddress();
+
+    /**
+     * Sets the {@link InetSocketAddress} remote peer provided by the proxy.
+     */
+    void setProxySourceAddress(InetSocketAddress addr);
```
I do not like this API change much as...

 - 1. We do not know that a null value might be returned. NPEs are waiting to happen
 - 2. Nothing forbids me to set the proxy source and not the destination, or vice versa.
 
 Let me suggest to...
 
  - Create a `ProxyInformation` POJO
  
```
class ProxyInformation {
    private final InetSocketAddress source;
    private final InetSocketAddress destination;
    
    // constructor, equals, hash code, getters...
```

Then in protocol session we can... 

```
    void setProxyInformation(ProxyInformation info);
    Optional<ProxyInformation> getProxyInformation();
```

No more null, no more partially specified proxy info.

WDYT ?

ouvtam replied:
Had the same thought first, but I wanted it to keep simple. I will do so.

chibenwa commented on server/apps/distributed-app/docs/modules/ROOT/pages/configure/smtp.adoc:
```diff
@@ -54,6 +54,12 @@ this case, if no body is present, the value "localhost" will be used.
+(https://www.haproxy.org/download/2.7/doc/proxy-protocol.txt) is used and might be compatible
+with other proxies (e.g. traefik). If enabled, it is *required* to initiate the connection
+using HAProxy's proxy protocol.
+
```
Can this be documented in imap.adoc too?

ouvtam replied:
Yes, we can copy it, but I did not test this for IMAP.

chibenwa reviewed (commented):
Thanks for the high quality contribution, which I believe is a great addition to James.

Regarding documentation we would benefit from mentionning the `proxyRequired` setting both in `src/site/xdoc/server/config-smtp-lmtp.xml` and `src/site/xdoc/server/config-imap4.xml`.

Also I would be super happy if we could set up a little exemple demoing the full set up. Let me suggest adding an exemple docker-compose, in `exemples/haproxy`, that starts a HAProxy and a James memory using the proxy protocol. using telnet with the HAproxy container, I would see the connection logs in the James container with the right source IP. If setting up such an exemple is complicated for you, I will be more than happy to help.

Best regards,

Benoit

ouvtam commented:
I will add an example for haproxy as you suggested and update this PR.

chibenwa commented on protocols/smtp/src/test/java/org/apache/james/protocols/smtp/utils/BaseFakeSMTPSession.java:
```diff
@@ -27,6 +27,7 @@
 
 import org.apache.james.core.Username;
 import org.apache.james.protocols.api.ProtocolSession;
+import org.apache.james.protocols.api.ProxyInformation;
```
Are we missing the ProxyInformation file in this pull request?

ouvtam replied:
Resolved.

ouvtam commented:
I added a sample docker setup to illustrate how to configure james with haproxy, have fun!

chibenwa commented on protocols/api/src/main/java/org/apache/james/protocols/api/ProxyInformation.java:
```diff
@@ -0,0 +1,54 @@
+
+
+    @Override
+    public final int hashCode() {
```
If we implement hashcode, we also need equals?

chibenwa commented on protocols/api/src/main/java/org/apache/james/protocols/api/ProxyInformation.java:
```diff
@@ -0,0 +1,54 @@
+    private final InetSocketAddress source;
+    private final InetSocketAddress destination;
+
+    public ProxyInformation(InetSocketAddress source, InetSocketAddress destination) {
+        this.source = source;
+        this.destination = destination;
+    }
```
```suggestion
    public ProxyInformation(InetSocketAddress source, InetSocketAddress destination) {
        Preconditions.checkNotNull(source);
        Preconditions.checkNotNull(destination);
        
        this.source = source;
        this.destination = destination;
    }
```

chibenwa reviewed (commented):
Can we move the example docker-compose in  this folder https://github.com/apache/james-project/tree/master/examples ?

chibenwa commented on server/apps/proxy-smtp/docker-compose.yml:
```diff
@@ -0,0 +1,33 @@
+version: '3.0'
+
+# Explanation:
+# - starts a haproxy server on address haproxy:25 that forwards connections to james:2525
+# - the service "helo" just sends a HELO to haproxy that will be forwarded to james
+# - james receives the origin ip address of the helo container provided by haproxy
+# - you can also use telnet to connect to haproxy if you expose haproxy's port 25
+#
+# Caveat: james might take a while to start. Thus start james first,
+# then haproxy and finally helo. Observe the log ouput of the helo
+# container to see if it works correctly.
```
I would be moving this explanation into a README.md file, which would make it easier to find and read using GitHub. WDYT ?

ouvtam replied:
Good idea.

ouvtam commented:
> Can we move the example docker-compose in this folder https://github.com/apache/james-project/tree/master/examples ?

Moved example and added proper README explaining the proxy use case.

chibenwa commented:
Hello,

I did get a closer look this morning:

 - Played with IMAP and Proxy protocol to make it work
 - Fix POJO bean contract for ProxyInformation
 - Refactor the Netty code to compose distinct handlers rather than keep current handlers continuing to grow out of control
 - Record proxy infos into the logging MDC
 - Re-indent test
 
 Hope it helps, and congrats for the good work.

ouvtam commented:
Superb! Thanks for your feedback and your solid refactor of the Netty handlers that allows the use in IMAP.

Arsnael commented on protocols/api/src/test/java/org/apache/james/protocols/api/ProxyInformationTest.java:
```diff
@@ -0,0 +1,30 @@
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/package org.apache.james.protocols.api;
```
the package declaration should be in the line under

Arsnael commented on protocols/netty/src/main/java/org/apache/james/protocols/netty/HAProxyMessageHandler.java:
```diff
@@ -0,0 +1,74 @@
+package org.apache.james.protocols.netty;
```
Missing licence

vttranlina commented on server/protocols/protocols-library/src/main/java/org/apache/james/protocols/lib/netty/AbstractConfigurableAsyncServer.java:
```diff
@@ -444,16 +451,23 @@ protected ChannelHandlerFactory getFrameHandlerFactory() {
     protected abstract ChannelInboundHandlerAdapter createCoreHandler();
-    
+
+    protected abstract ChannelInboundHandlerAdapter createProxyHandler();
```
I don't find any where use it
